server/hotrod/src/main/scala/org/infinispan/server/hotrod/Decoder2x.scala

```diff
@@ -397,7 +419,28 @@ object Decoder2x extends AbstractVersionedDecoder with ServerConstants with Log
       }
    }
 
-   override def customReadValue(header: HotRodHeader, buffer: ByteBuf, cache: Cache): AnyRef = null
+   override def customReadValue(decoder: HotRodDecoder, h: HotRodHeader,
+       hrCtx: CacheDecodeContext, buffer: ByteBuf, cache: Cache): AnyRef = {
+      h.op match {
+         case PutAllRequest =>
+            var map = hrCtx.putAllMap
+            if (map == null) {
+              map = new HashMap[Bytes, Bytes]
+              hrCtx.putAllMap = map
+            }
+            for (i <- map.size until hrCtx.params.valueLength) {
+              val k = readRangedBytes(buffer)
+              val v = readRangedBytes(buffer)
+              map.put(k, v)
+              // We check point after each read entry
+              decoder.checkpoint
```

Comment:
> The most important thing about this PR is making sure after key and value a checkpoint was set in case it had to be re-read as a result of not having all the information. I wasn't aware that you could just call checkpoint without passing in a decoder state...

Yeah when there is no state provided it keeps it in the same state but only moves the data ahead that far as you have read.  You can actually do a replaying decoder without an enum state as well.  I retain the state in the map on the decoder context of previously read entries and then checkpoint to mark I read up to those bytes.

> From what I understand, if not enough bytes were found, an exception is thrown and then when you read again, you start with the decode() callback. That's why we added the different state enums (DECODE_HEADER) and the match/case statement at the top. Hence, I don't know how your code can work really... because once you go to the top of the decode() and you want to read the key/value pair at which it stopped, you still need to go through the entire call stacktrace all the way to here...

It actually throws a special Error instance that they hold for this special purpose (reused so only a single stack trace is ever generated).  If you notice I got through each step of the decode phase indivually (this is I how I expected all commands to work).

> My plan on how to implement this involved adding a new state e.g. HotRodDecoderState.DECODE_KEY_VALUE_PAIR and cache the number of params read so far, including total number of pairs to read, and number of pairs read so far, and when finding that checkpoint state, calculate how many are left to read and loop for that many.... but maybe your approach works already, which if it did, it'd be simpler.

I do cache the number of ~~params~~ entries read so far in the map on the decoder context.  Then in the for loop it starts at that number until it finds the appropriate number of entries.

> A quick way to test your approach really works (if you've not done so already!) is to use very big values and see how it behaves, whether it fails to read all the bytes for one of the key/value pairs and see if it works as expected.

I have tested it and seen in the debugger, it usually has to replay on most of the tests.